Declarative duts_config and DUT collection refactor

Deixx opened the pull request:
On editing duts_config replace configured duts
instead of adding all duts from modified config.
This allows for both adding and removing duts.
Current behavior led to duplicates if config contained
old duts in addition to new ones.

Deixx committed:
Reformatting

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
Declarative duts_config

On editing duts_config replace configured duts
instead of adding all duts from modified config.
This allows for both adding and removing duts.
Current behavior led to duplicates if config contained
old duts in addition to new ones.

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

mmichal10 commented on runnerd:
```diff
@@ -121,38 +121,51 @@ class DutsManager:
+            free_dut = self.duts_queue.get()
+            if not dut_filter or dut_filter(free_dut):
+                dut = free_dut
                 break
```
`dut_filter` is never used. Please consider removing it

Deixx committed:
Remove unused filtering + change DUT reconfiguration

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
DUTs collecting and logging improvements

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

mmichal10 commented on runnerd:
```diff
@@ -140,33 +140,27 @@ class DutsManager:
+
+    def get_free_dut(self):
         dut = None
         for _ in self.duts:
```
`duts_queue.get()` won't timeout, so the loop is redundant

Deixx replied:
Changed get to non-blocking

Deixx committed:
More changes to collecting DUTs

Handle FileNotFoundError (when overwriting config file)
Look for DUT only if queue is not empty

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

mmichal10 commented on runnerd:
```diff
@@ -122,28 +124,39 @@ class DutsManager:
     base_dut_config_path = "configs/base_dut_config.yml"
 
     def __init__(self, duts_config_file):
+        self.lock = Lock()
```
Could you use more descriptive name to explain what is the lock supposed to protect?

mmichal10 commented on runnerd:
```diff
@@ -133,9 +133,15 @@ class DutsManager:
+                    else:
+                        break
+                except FileNotFoundError:
+                    pass
```
runner might get stuck on this loop without providing any information to the user. How about adding a log and printing it every few seconds?

mmichal10 commented on runnerd:
```diff
@@ -122,28 +124,39 @@ class DutsManager:
-                self.mark_free(self.duts[ip])
+            self.duts_config = self.duts_config_file.load()
+            configured_duts = self.duts_config.get('duts', [])
+            if configured_duts is None:
```
This condition is never true

Deixx replied:
You'd be surprised ;)
Below are example results for different configs:
```
duts:
  - dut1
```
\- gives non-empty list 

```duts: []``` - gives empty list

```duts:```  - gives None

mmichal10 commented on runnerd:
```diff
@@ -133,9 +133,15 @@ class DutsManager:
 
     def collect_duts(self):
         with self.lock:
```
You could acquire the lock only after the loop

Deixx committed:
Refactor

Removed redundant code, clarified lock, added logging for
missing DUT config file

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
Prevent scheduling tests on the same DUT

Check if DUT is in use before adding it to queue
Prevent duplicated entries in queue
Handle faulty config load

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

pep8speaks commented:
Hello @Deixx! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`common.py`](https://github.com/Open-CAS/test-runner/blob/86ebcd240dbe356ec54679ff87344fa2486bc674/common.py):

> [Line 41:80](https://github.com/Open-CAS/test-runner/blob/86ebcd240dbe356ec54679ff87344fa2486bc674/common.py#L41): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

Deixx committed:
Changed the order of looking for tests and DUTs

Acquire test and then look for DUT to run it on

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
Collect DUTs in a separate thread

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
Simplify getting DUTs

With separate collection thread we can get rid of double loop.

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>

Deixx committed:
Unify termination behaviour

Daemonized and non-daemonized runner will now behave the same
after receiving a SIGTERM or SIGINT signal and close all
the processes in the process group. The behaviour was different
also for Ctrl+C (affected process group) and sending SIGINT
through kill command (affected only the main process).

Signed-off-by: Daniel Madej <daniel.madej@huawei.com>
Signed-off-by: Robert Baldyga <robert.baldyga@unvertical.com>
